scripts/travis/build.sh

```diff
@@ -49,6 +49,34 @@ else
   EXTRA_ARGS=''
 fi
 
+# For PR builds, Skip module-specific tests if its module dependencies haven't been touched
+if [ "$TRAVIS_PULL_REQUEST" != 'false' ]; then
+  CONDITIONAL_TESTING_MODULES='d2 r2-int-test restli-int-test'
+  echo "This is a PR build, so conditional testing will be done for these subprojects: [${CONDITIONAL_TESTING_MODULES// /,}]"
+  # If any Gradle file was touched, run all tests just to be safe
+  if (git diff ${TRAVIS_BRANCH}...HEAD --name-only | grep '\.gradle' > /dev/null); then
+    echo "This PR touches a file matching *.gradle, so tests will be run for all subprojects."
+  else
+    # Have to prime the comma-separated list with a dummy value because list construction in bash is hard...
+    EXTRA_ARGS="${EXTRA_ARGS} -Ppegasus.skipTestsForSubprojects=primer"
+    # For all the following modules (which have lengthy tests), determine if they can be skipped
+    for MODULE in $CONDITIONAL_TESTING_MODULES; do
+      echo ''
+      echo "Checking test dependencies for subproject $MODULE..."
+      MODULE_DEPENDENCIES="$(./scripts/get-module-dependencies $MODULE testRuntimeClasspath | tr '\n' ' ')"
+      echo "Computed as [${MODULE_DEPENDENCIES// /,}]"
+      PATH_MATCHING_REGEX="^\\($(echo $MODULE_DEPENDENCIES | sed -z 's/ /,/g;s/,$/\n/;s/,/\\|/g')\\)/"
```

Comment:
why do you need the two step process? replace space with comma and then replace comma with pipe (|) ?

Reply:
It's 3 steps:
1. Replace spaces with comma.
2. Delete trailing comma.
3. Replace commas with escaped pipe `\|`

For some reason, I thought this was cleaner, but evidently not so much if it confuses you. I'll correct this.